src/Swift.Runtime/src/Metadata/TypeMetadata.cs

```diff
@@ -249,4 +253,113 @@ public override int GetHashCode ()
+    /// <returns>true on success false otherwise</returns>
+    public static bool TryGetTypeMetadata<T>([NotNullWhen(true)] out TypeMetadata? result)
+    {
+        if (cache.TryGet(typeof(T), out result))
+            return true;
+        return TryGetTypeMetadataUncached<T>(out result);
+    }
+
+    /// <summary>
+    /// Attempt to get the Swift type metadata for the given type
+    /// </summary>
+    /// <typeparam name="T">The type of the object</typeparam>
+    /// <returns>The result of the looked up type metadata on success</returns>
+    /// <exception cref="Exception">Throws when lookup fails</exception>
+    public static TypeMetadata GetTypeMetadataOrThrow<T>()
+    {
+        if (TryGetTypeMetadata<T>(out var result))
+            return result.Value;
+        throw new Exception(string.Format("Unable to get type metadata for type {0}", typeof(T).Name));
+    }
+
+    struct MetadataHelper<T> where T: ISwiftObject
+    {
+        public static TypeMetadata GetTypeMetadata()
+        {
+            return T.GetTypeMetadata();
+        }
+    }
+
+    /// <summary>
+    /// Attempt to get the Swift type metadata but without accessing the cache
+    /// </summary>
+    /// <typeparam name="T">The type of the object</typeparam>
+    /// <param name="result">The result of looked up type metadata</param>
+    /// <returns>true on success false otherwise</returns>
+    /// <exception cref="NotImplementedException">Throws if unable to look up the ISwiftObject.GetTypeMetadata method.</exception>
+    static bool TryGetTypeMetadataUncached<T>([NotNullWhen(true)] out TypeMetadata? result)
+    {
+        var type = typeof(T);
+        if (typeof(ISwiftObject).IsAssignableFrom(type))
+        {
+            var helper = typeof(MetadataHelper<>).MakeGenericType(type);
+            result = (TypeMetadata)helper.GetMethod("GetTypeMetadata")!.Invoke(null, null)!;
+            return true;
+        }
+
+        // NB - all further methods here should finish by putting the type into the cache
+
+        // TODO: handle tuples https://github.com/dotnet/runtimelab/issues/2873
+
+        // TODO: handle closures https://github.com/dotnet/runtimelab/issues/2874
+
+        // TODO: handle existential containers https://github.com/dotnet/runtimelab/issues/2875
+
+        result = null;
+        return false;
+    }
+
+    /// <summary>
+    /// Returns an enumeration of known Type and TypeMetadata objects
+    /// </summary>
+    /// <returns>An enumeration of known Type and TypeMetadata objects</returns>
+    /// <exception cref="Exception">Throws if unable to load a library containing known types</exception>
+    static IEnumerable<(Type, TypeMetadata)> KnownMetadata()
+    {
+        var libraryHandle = NativeLibrary.Load(KnownLibraries.SwiftCore);
+        if (libraryHandle == IntPtr.Zero)
+            throw new Exception(string.Format("Unable to load library {0}", KnownLibraries.SwiftCore));
+        // types from libSwiftCore
```

Comment:
All these mangled names should likely be consolidated into a single place and documented on how to retrieve them from `libSwiftCore`.